src/mscorlib/src/System/String.Comparison.cs

```diff
@@ -0,0 +1,1160 @@
+        {
+            if (strA == null || strB == null)
+            {
+                if (object.ReferenceEquals(strA, strB))
+                {
+                    // They're both null
+                    return 0;
+                }
+
+                return strA == null ? -1 : 1;
+            }
+
+            // COMPAT: Checking for nulls should become before the arguments are validated,
+            // but other optimizations which allow us to return early should come after.
+
+            if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException("length", Environment.GetResourceString("ArgumentOutOfRange_NegativeCount"));
+            }
+
+            if (indexA < 0 || indexB < 0)
+            {
+                string paramName = indexA < 0 ? "indexA" : "indexB";
+                throw new ArgumentOutOfRangeException(paramName, Environment.GetResourceString("ArgumentOutOfRange_Index"));
+            }
+            
+            int lengthA = Math.Min(length, strA.Length - indexA);
+            int lengthB = Math.Min(length, strB.Length - indexB);
+
+            if (lengthA < 0 || lengthB < 0)
+            {
+                string paramName = lengthA < 0 ? "indexA" : "indexB";
+                throw new ArgumentOutOfRangeException(paramName, Environment.GetResourceString("ArgumentOutOfRange_Index"));
+            }
+
+            if (length == 0 || (object.ReferenceEquals(strA, strB) && indexA == indexB))
+            {
+                return 0;
+            }
+
+            return CompareOrdinalHelper(strA, indexA, lengthA, strB, indexB, lengthB);
+        }
+
+        // Compares this String to another String (cast as object), returning an integer that
+        // indicates the relationship. This method returns a value less than 0 if this is less than value, 0
+        // if this is equal to value, or a value greater than 0 if this is greater than value.
+        //
+        [Pure]
+        public int CompareTo(Object value)
+        {
+            if (value == null)
+            {
+                return 1;
+            }
+
+            string other = value as string;
+
+            if (other == null)
+            {
+                throw new ArgumentException(Environment.GetResourceString("Arg_MustBeString"));
+            }
+
+            return CompareTo(other); // will call the string-based overload
+        }
+    
+        // Determines the sorting relation of StrB to the current instance.
+        //
+        [Pure]
+        public int CompareTo(String strB)
+        {
+            return string.Compare(this, strB, StringComparison.CurrentCulture);
+        }
+
+        // Determines whether a specified string is a suffix of the the current instance.
+        //
+        // The case-sensitive and culture-sensitive option is set by options,
+        // and the default culture is used.
+        //        
+        [Pure]
+        public Boolean EndsWith(String value) {
```

Comment:
@justinvp Thanks for reviewing. I had originally proposed putting the `*With` methods next to `Contains` / `IndexOf` in https://github.com/dotnet/corert/issues/906#issuecomment-243200109, but I decided not to for a couple of reasons:

- The functions in `String.Searching.cs` are checking different portions of the source string to see if they're equal to something else. The functions in `String.Comparison.cs` test a specific substring of the source string to see if they're equal to something else. The `*With` functions most closely fit the latter definition.

- `s.EndsWith(value)` is equivalent to `s.Substring(s.Length - value.Length) == value` or `string.Compare(s, s.Length - value.Length, value, 0, value.Length) == 0`, whereas it can't be put in terms of Contains/IndexOf.

Sorry for the confusion.